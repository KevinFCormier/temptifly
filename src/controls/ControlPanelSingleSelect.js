'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {
  FormGroup,
  Popover,
  Select,
  SelectOption,
  SelectVariant,
  Spinner } from '@patternfly/react-core'
import HelpIcon from '@patternfly/react-icons/dist/js/icons/help-icon'
import _ from 'lodash'

class ControlPanelSingleSelect extends React.Component {
  static propTypes = {
    control: PropTypes.object,
    controlId: PropTypes.string,
    handleChange: PropTypes.func,
    i18n: PropTypes.func
  };

  constructor(props) {
    super(props)
    this.state = {
      open: false
    }
  }

  setControlRef = (control, ref) => {
    control.ref = ref
  };

  render() {
    const { open } = this.state
    const { controlId, i18n, control, handleChange } = this.props
    const {
      name,
      placeholder = '',
      available = [],
      validation = {},
      exception,
      tooltip,
      disabled,
      isLoading,
      isFailed
    } = control
    let { active } = control
    if (!active) {
      if (isLoading) {
        active = i18n(
          _.get(control, 'fetchAvailable.loadingDesc', 'resource.loading'))
      } else if (isFailed) {
        active = i18n('resource.error')
      } else if (available.length === 0) {
        active = i18n(
          _.get(control, 'fetchAvailable.emptyDesc', 'resource.none'))
      }
    }
    const setOpen = (open) => {
      this.setState({open})
    }
    const onChange = (value) => {
      control.active = value
      handleChange()
    }
    const validated = exception ? 'error' : undefined
    return (
      <React.Fragment>
        <div
          className="creation-view-controls-singleselect"
          ref={this.setControlRef.bind(this, control)}
        >
          <FormGroup
            id={`${controlId}-label`}
            label={name}
            isRequired={validation.required}
            fieldId={controlId}
            helperTextInvalid={exception}
            validated={validated}
            labelIcon={
              /* istanbul ignore next */
              tooltip ? (
                <Popover
                  id={`${controlId}-label-help-popover`}
                  bodyContent={tooltip}
                >
                  <button
                    id={`${controlId}-label-help-button`}
                    aria-label="More info"
                    onClick={(e) => e.preventDefault()}
                    className="pf-c-form__group-label-help"
                  >
                    <HelpIcon noVerticalAlign />
                  </button>
                </Popover>
              ) : (
                <React.Fragment />
              )
            }
          >
            {isLoading ? (
              <div className="creation-view-controls-singleselect-loading">
                <Spinner size="md" />
                <div>{active}</div>
              </div>
            ) : (
              <Select
                aria-labelledby={`${controlId}-label`}
                spellCheck={false}
                isOpen={open}
                onToggle={() => {setOpen(!open)}}
                variant={SelectVariant.typeahead}
                onSelect={(_event, value) => {
                  onChange(value)
                  setOpen(false)
                }}
                selections={active}
                onClear={() => {onChange(undefined)}}
                placeholderText={placeholder}
                isDisabled={disabled}
              >
                {available.map((item, inx)=>{
                  /* eslint-disable-next-line react/no-array-index-key */
                  return <SelectOption key={inx} value={item} />
                })}
              </Select>
            )}
            {validated === 'error' ? (
              <div style={{ borderTop: '1.75px solid red', paddingBottom: '6px', maxWidth: '800px' }}></div>
            ) : (
              <React.Fragment />
            )}
          </FormGroup>
        </div>
      </React.Fragment>
    )
  }
}

export default ControlPanelSingleSelect
