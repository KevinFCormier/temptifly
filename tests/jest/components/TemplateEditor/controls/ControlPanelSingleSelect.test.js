'use strict'

import React from 'react'
import ControlPanelSingleSelect from '../../../../../src/controls/ControlPanelSingleSelect'
import renderer from 'react-test-renderer'

import { control } from '../../TestingData'

describe('ControlPanelSingleSelect component', () => {
  it('renders as expected', () => {
    const fn = jest.fn()
    const component = renderer.create(
      <ControlPanelSingleSelect
        key={'key'}
        control={control}
        controlId={'controlId'}
        handleChange={fn}
        i18n={fn}
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
