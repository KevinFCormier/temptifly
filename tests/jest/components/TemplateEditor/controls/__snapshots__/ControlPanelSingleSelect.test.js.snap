// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanelSingleSelect component renders as expected 1`] = `
<div
  className="creation-view-controls-singleselect"
>
  <div
    className="pf-c-form__group"
    id="controlId-label"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="controlId"
      >
        <span
          className="pf-c-form__label-text"
        >
          creation.app.name
        </span>
      </label>
       
      <button
        aria-label="More info"
        className="pf-c-form__group-label-help"
        id="controlId-label-help-button"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          aria-labelledby={null}
          fill="currentColor"
          height="1em"
          role="img"
          style={null}
          viewBox="0 64 1024 1024"
          width="1em"
        >
          <path
            d="M512.059-73.143c-282.338 0-512.059 229.673-512.059 512.025 0 282.235 229.721 511.975 512.059 511.975 282.281 0 511.941-229.735 511.941-511.975 0.005-282.352-229.659-512.025-511.941-512.025zM512.059 826.523c-213.826 0-387.728-173.856-387.728-387.643 0-213.89 173.904-387.694 387.728-387.694 213.717 0 387.671 173.803 387.671 387.694 0.005 213.785-173.957 387.643-387.671 387.643zM585.143 164.571v109.714c0 4.951-1.808 9.237-5.429 12.857s-7.906 5.429-12.857 5.429h-109.714c-4.953 0-9.239-1.808-12.857-5.429s-5.429-7.906-5.429-12.857v-109.714c0-4.951 1.81-9.237 5.429-12.857s7.904-5.429 12.857-5.429h109.714c4.951 0 9.237 1.808 12.857 5.429s5.429 7.906 5.429 12.857zM521.616 365.714c118.343 0 214.286 88.965 214.286 187.429s-95.943 178.286-214.286 178.286c-173.045 0-208.091-93.714-213.963-171.113 0-7.936 6.729-9.838 15.872-9.838s108.073 0 113.143 0c6.096 0 14.475 0.633 17.554 10.571 0 48.857 135.968 54.953 135.968-7.904 0-31.506-29.717-63.817-68.571-66.194s-82.286-7.607-82.286-54.199c0-13.022 0-25.673 0-44.693 0-19.015 9.717-22.343 27.431-22.343s54.853-0.002 54.853-0.002z"
            transform="rotate(180 0 512) scale(-1 1)"
          />
        </svg>
      </button>
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <div
        className="pf-c-select"
        data-ouia-component-id="OUIA-Generated-Select-typeahead-1"
        data-ouia-component-type="PF4/Select"
        data-ouia-safe={true}
      >
        <div
          className="pf-c-select__toggle pf-m-typeahead"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          <div
            className="pf-c-select__toggle-wrapper"
          >
            <input
              aria-activedescendant={null}
              aria-label=""
              autoComplete="off"
              className="pf-c-form-control pf-c-select__toggle-typeahead"
              disabled={false}
              id="pf-select-toggle-id-0-select-typeahead"
              onChange={[Function]}
              onClick={[Function]}
              placeholder=""
              type="text"
              value="test"
            />
          </div>
          <button
            aria-label="Clear all"
            className="pf-c-button pf-m-plain pf-c-select__toggle-clear"
            disabled={false}
            onClick={[Function]}
            onKeyDown={[Function]}
            type="button"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
              />
            </svg>
          </button>
          <button
            aria-expanded={false}
            aria-haspopup="listbox"
            aria-label="Options menu"
            aria-labelledby="controlId-label pf-select-toggle-id-0"
            className="pf-c-button pf-c-select__toggle-button pf-m-plain"
            disabled={false}
            id="pf-select-toggle-id-0"
            onClick={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-c-select__toggle-arrow"
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              />
            </svg>
          </button>
        </div>
      </div>
      
    </div>
  </div>
</div>
`;
